### ISSUE TITLE
Port tap stream classes without channel parameter

### ISSUE THREAD

**Issue by lmmx** [lmmx/beeb#11]
(Towards #4: porting tap stream pull to beeb)

- Porting the functionality will potentially leave tap without a working generic download functionality of its own, but it never really had that (and maybe making that explicit will enable it to get it)

**Comment by lmmx** [comment]
- Stream URL set handling ported in d23b0b45aac30e8532744a8db4fe7cd2fa26a8dc
- Stream URL set fetching ported in 820b8f9d89111dcf6d621739e757523c6f04655e

I don't think more can be ported without splitting the class (i.e. making a base `Stream` in beeb that's subclassed in tap)

Since beeb is imported into tap, it doesn't make sense for tap to be imported into beeb, so beeb must supply the base class

Change tap's Stream to `TapStream` and figure out how to use a classmethod to create one from a `beeb.Stream`... or something like this?

**Comment by lmmx** [comment]
The signature for tap’s class `Stream.__init__` is

```py
class Stream(Episode):
    def __init__(
        self,
        broadcaster,
        station,
        programme,
        date,
        urlset,
        transcribe=False,
        reload=False,
        load_full_transcripts=True,
        **preproc_opts,
    )
```

Note that It inherits from `tap.stream.Episode`.

If beeb is to supply a base class then it must become dual inheritance...?

- (Hold on, how will that work when I want to support other broadcasters?)

```py
class Stream(beeb.Stream, Episode):
```

and its __init__ method could simply call `super().__init__`?

The signature for a simple base class for BBC streams would look more like this:

```py
class BeebStream:
    def __init__(
        self,
        station,
        programme,
        date,
        urlset,
        reload=False,
)
```

The transcription-specific features would not be included, only the `reload` would be used to avoid pulling the stream from its canonical location.

Note that this must **not** be package specific, else `tap` sharing the stream mechanism of `beeb` would result in the streams downloaded in `tap` being saved under the `beeb` directory!

- **Edit** achieved by setting `_root_store_dir` as an overridable property on which the other download directory properties (`_store_dir`) are determined at runtime, i.e. overriding it to be `tap.data.store.broadcasters._dir_path` will switch the download directory to the tap directory rather than beeb

Removing the above options leaves:

```py
class TapStream(Episode):
    def __init__(
        self,
        transcribe=False,
        load_full_transcripts=True,
        **preproc_opts,
    )
```

which... seems much more manageable?

Maybe a generic class could be made for anything else?

TODO: figure out how to handle these different classes rather than a single canonical `tap.Stream` class

- Perhaps it’s not sensible to just make a ‘stream’ a narrower type of ‘episode’ (i.e. split the object rather than just use hierarchy to keep specifying details upon it)

I would suggest using the `TapStream` as a mixin with the `BeebStream`, along the lines of:

in beeb:

```py
class BeebStream:
    def __init__(self,  station, programme, date, urlset, reload=False):
        ...
```

and in tap, this can be achieved with memoised dynamic inheritance:

```py
from beeb.stream import Stream as BeebStreamBase

class memoise:
    def __init__(self, f):
        self.f = f
        self.memo = {}

    def __call__(self, *args):
        return self.memo.setdefault(args, self.f(*args))

class TranscribeStreamMixIn(Episode):
    def __init__(self,  transcribe=False, load_full_transcripts=True, **preproc_opts):
        ...

class OtherStreamBase:
    def __init__(self, *args, **kwargs):
        raise NotImplementedError("Only BBC is currently supported")

@memoise
def make_stream(base_mixer):

    class Stream(TranscribeStreamMixIn, base_mixer):
        super().__init__(...) # call TranscribeStreamMixIn.__init__
        super(TranscribeStreamMixIn, self).__init__(...) # call base_mixer.__init__

    return Stream

a1 = make_stream(BBCMixer)
a2 = make_stream(BBCMixer)

print(a1 is a2) # True
```

Put simply: can greatly simplify the Stream object given PID only and assumption broadcaster is BBC - station is secondary importance. Rather than specify methods/properties like Station.beeb_channel_id just keep it simple and BBC-specific and subclass for tap.

Can't have "God class" TapStreamMixIn ∵ underlying mixins depend on init values: can't set mixins at runtime (at init time) but can use dynamic inheritance (and a memoised decorator prevents creating distinct classes from repeated calls with the same base).

Could make a StreamCreator class but I prefer the get_stream procedural approach (which is a class after the decorating memoise class anyway)

**Comment by lmmx** [comment]
Begun making breaking commits to tap: https://github.com/lmmx/tap/commit/90178fa8daf76621f201e7bb95ea38066d06107c
- `tap.stream.Stream` modelled on [`pathlib.Path`](https://github.com/python/cpython/blob/c8b5738810516df5722caf049003e9b319427bec/Lib/pathlib.py#L1071)

**Comment by lmmx** [comment]
Ported in 04e061e93d4135a0df2f6812a6b5daa3920eaba0
